File: zero-subarray/solution1.clj
(defn- get-subarrays
  [data]
  (loop [[head & tail] data
         intermediate-vec []
         intermediate-sum 0
         res []]
    (if head
      (let [intermediate-sum (+ intermediate-sum head)
            intermediate-vec (conj intermediate-vec head)]
        (if (= intermediate-sum 0)
          (recur tail intermediate-vec intermediate-sum (conj res intermediate-vec))
          (recur tail intermediate-vec intermediate-sum res)))
      res)))

(defn subarray
  [numbers]
  (->> numbers
       (iterate rest)
       (take-while seq)
       (mapcat get-subarrays)
       (filter seq)
       (into #{})))

;;
;; TESTS
;;

(let [data [3, 4, -7, 3, 1, 3, 1, -4, -2, -2]
      res #{[3, 4, -7],
            [4, -7, 3],
            [-7, 3, 1, 3],
            [3, 1, -4],
            [3, 1, 3, 1, -4, -2, -2],
            [3, 4, -7, 3, 1, 3, 1, -4, -2, -2]}
      res* (subarray data)]
  (when (not= res res*)
    (throw (Exception. (format "Expected %s got %s" res res*)))))

File: zero-subarray/solution1.py
#
# O(n^2) solution
#

def subarray(data):
  l = len(data)
  res = set()

  for i in range(0, l):
    sofar = 0
    for j in range(i, l):
      sofar += data[j]
      if sofar == 0:
        res.add(tuple(data[i:j+1]))
  
  return res

# 
# TESTS
# 

data = [3, 4, -7, 3, 1, 3, 1, -4, -2, -2]

res = {(3, 4, -7), 
       (4, -7, 3),
       (-7, 3, 1, 3),
       (3, 1, -4),
       (3, 1, 3, 1, -4, -2, -2),
       (3, 4, -7, 3, 1, 3, 1, -4, -2, -2)}

res_ = subarray(data)
if res_ != res:
  raise(Exception("Expected %s, got %s" % (res, res_)))